tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php

```diff
@@ -402,10 +407,20 @@ public function testResultCompatibilityWhenExecutingCacheQuery(): void
             $this->connection->getDatabasePlatform()->getDummySelectSQL(),
             [],
             [],
-            new QueryCacheProfile(1, 'cacheKey', new ArrayCache())
+            new QueryCacheProfile(1, 'cacheKey', $this->getArrayCache())
         );
 
         self::assertInstanceOf(Result::class, $result);
         self::assertInstanceOf(Driver\ResultStatement::class, $result);
     }
+
+    private function getArrayCache(): Cache
+    {
+        $cache = class_exists(DoctrineProvider::class)
+            ? DoctrineProvider::wrap(new ArrayAdapter())
+            : new ArrayCache();
+        assert($cache instanceof Cache);
```

Comment:
Why is this assertion necessary? Shouldn't both versions implement the expected interface?

Reply:
Yes, but when Doctrine Cache 2 is installed, `ArrayCache` is undefined and Psalm is confused. We can remove the assertion on 3.2.x.

Reply:
Shouldn't there be also a condition like `class_exists()`? From the static analysis PoV, the fact that the `DoctrineProvider` class doesn't exist doesn't mean that `ArrayCache` does.

So it could be something like:
```php
if (class_exists(DoctrineProvider::class)) {
    return DoctrineProvider::wrap(new ArrayAdapter());
}

if (class_exists(ArrayCache::class)) {
    return new ArrayCache();
}

self::markTestSkipped();
```

Reply:
I'm using your code now. Psalm seems to be less confused that way. 🙂